src/connection/network/proto/logic.rs

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+use typedb_protocol::Rule as RuleProto;
+use typeql_lang::{
+    parse_pattern, parse_variable,
+    pattern::{Pattern, Variable},
+};
+
+use super::{IntoProto, TryFromProto};
+use crate::{common::Result, Error, Rule};
+
+impl TryFromProto<RuleProto> for Rule {
+    fn try_from_proto(proto: RuleProto) -> Result<Self> {
+        let when = match parse_pattern(&proto.when) {
+            Ok(Pattern::Conjunction(conjunction)) => conjunction,
+            Ok(other) => return Err(Error::Other(format!("When parse error: {other:?}"))),
+            Err(error) => return Err(Error::Other(format!("{error:?}"))),
```

Comment:
If there isn't a TypeQL error we can use here, we should create a new error enum in `common/error` and use that.